stub/src/main/java/io/grpc/stub/ServerCalls.java

```diff
@@ -348,7 +349,13 @@ public void onNext(RespT response) {
         call.sendHeaders(new Metadata());
         sentHeaders = true;
       }
-      call.sendMessage(response);
+
+      if (!call.getMethodDescriptor().getType().serverSendsOneMessage()) {
+        call.sendMessage(response);
+      } else {
+        // delay the sendMessage() to onComplete()/onError()
+        unaryResponse = response;
```

Comment:
IMO this should throw an exception if the server sends multiple responses/overwrites the previous obe.